prusti-viper/src/encoder/spec_function_encoder.rs

```diff
@@ -0,0 +1,156 @@
+        let pre_name = self.encoder.encode_spec_func_name(self.procedure.get_id(),
+                                                          SpecFunctionKind::Pre);
+        let post_name = self.encoder.encode_spec_func_name(self.procedure.get_id(),
+                                                           SpecFunctionKind::Post);
+
+        if let None = self.contract {
+            return Ok(Vec::new());
+        }
+
+        let contract = compute_procedure_contract(
+            self.proc_def_id,
+            self.encoder.env().tcx(),
+            self.contract.clone().unwrap(),
+            Some(&self.encoder.current_tymap())
+        ).to_def_site_contract();
+
+        let pre_func = self.encode_pre_spec_func(&contract);
+        let post_func = self.encode_post_spec_func(&contract);
+
+        Ok(vec! [pre_func, post_func])
+    }
+
+    fn encode_pre_spec_func(&self, contract: &ProcedureContract<'tcx>) -> vir::Function {
+        let mut func_spec: Vec<vir::Expr> = vec![];
+
+        let encoded_args: Vec<vir::LocalVar> = contract
+            .args
+            .iter()
+            .map(|local| self.encode_local(local.clone().into()).into())
+            .collect();
+
+        for item in contract.functional_precondition() {
+            func_spec.push(self.encoder.encode_assertion(
+                &item,
+                &self.mir,
+                Some(&""),
+                &encoded_args.iter().map(|e| -> vir::Expr { e.into() }).collect::<Vec<_>>(),
+                None,
+                true,
+                None,
+                ErrorCtxt::GenericExpression,
+            ));
+        }
+
+        vir::Function {
+            name: self.encoder.encode_spec_func_name(self.procedure.get_id(),
+                                                     SpecFunctionKind::Pre),
+            formal_args: encoded_args.into_iter()
+                                     .skip(1)
+                                     .collect(), // skip "self" for now
+            return_type: vir::Type::Bool,
+            pres: Vec::new(),
+            posts: Vec::new(),
+            body: Some(func_spec.into_iter()
+                                .map(|post| SnapshotSpecPatcher::new(self.encoder).patch_spec(post))
+                                .conjoin()),
+        }
+    }
+
+    fn encode_post_spec_func(&self, contract: &ProcedureContract<'tcx>) -> vir::Function {
+        let mut func_spec: Vec<vir::Expr> = vec![];
+
+        let mut encoded_args: Vec<vir::LocalVar> = contract
+            .args
+            .iter()
+            .map(|local| self.encode_local(local.clone().into()).into())
+            .collect();
+        let encoded_return = self.encode_local(contract.returned_value.clone().into());
+
+        for item in contract.functional_postcondition() {
+            func_spec.push(self.encoder.encode_assertion(
+                &item,
+                &self.mir,
+                Some(&""),
```

Comment:
`None`

Reply:
I think you somehow reviewed an old version? The `None`s were added in cb3591ee0bc4616336223143d9a88c166c40ca63.